src/brpc/policy/mysql_protocol.cpp

```diff
@@ -0,0 +1,286 @@
+            std::string auth_cmd, params_cmd;
+            std::string user, password, schema, params;
+            MysqlParseDSN(ctx->user(), &user, &password, &schema, &params);
+            LOG(INFO) << user;
+            LOG(INFO) << password;
+            LOG(INFO) << schema;
+            LOG(INFO) << params;
+            if (MysqlPackAuthenticator(reply.auth(), user, password, schema, &auth_cmd) != 0 ||
+                MysqlPackParams(params, &params_cmd) != 0) {
+                LOG(ERROR) << "[MYSQL PARSE] wrong pack authentication data";
+                parseCode = PARSE_ERROR_ABSOLUTELY_WRONG;
+            } else {
+                butil::IOBuf auth_resp;
+                auth_resp.append(auth_cmd);
+                auth_resp.append(params_cmd);
+                LOG(INFO) << "auth_cmd.size=" << auth_cmd.size();
+                LOG(INFO) << "params_cmd.size=" << params_cmd.size();
+                auth_resp.cut_into_file_descriptor(socket->fd());
+            }
+        } else if (reply_size > 0) {
+            size_t i;
+            for (i = 0; i < reply_size; ++i) {
+                const MysqlReply& reply = msg->response.reply(i);
+                if (!reply.is_ok()) {
+                    break;
+                }
+            }
+            if (i < reply_size) {
+                LOG(ERROR) << msg->response;
+                parseCode = PARSE_ERROR_NO_RESOURCE;
+            } else {
+                butil::IOBuf raw_req;
+                raw_req.append(ctx->starter());
+                raw_req.cut_into_file_descriptor(socket->fd());
+                pi->with_auth = false;
+            }
+        } else {
+            LOG(ERROR) << "[MYSQL PARSE] wrong authentication step";
+            parseCode = PARSE_ERROR_ABSOLUTELY_WRONG;
+        }
+
+        // else if (reply.is_ok()) {
+        // } else if (reply.is_error()) {
+        //     LOG(ERROR) << reply;
+        //     parseCode = PARSE_ERROR_NO_RESOURCE;
+        // } else {
+        //     LOG(ERROR) << "[MYSQL PARSE] wrong authentication step";
+        //     parseCode = PARSE_ERROR_ABSOLUTELY_WRONG;
+        // }
+    }
+
+    if (bthread_id_unlock(cid) != 0) {
+        LOG(ERROR) << "[MYSQL PARSE] fail to unlock controller";
+        parseCode = PARSE_ERROR_ABSOLUTELY_WRONG;
+    }
+    return parseCode;
+}
+// "Message" = "Response" as we only implement the client for mysql.
+ParseResult ParseMysqlMessage(butil::IOBuf* source,
+                              Socket* socket,
+                              bool /*read_eof*/,
+                              const void* /*arg*/) {
+    if (source->empty()) {
+        return MakeParseError(PARSE_ERROR_NOT_ENOUGH_DATA);
+    }
+
+    PipelinedInfo pi;
+    if (!socket->PopPipelinedInfo(&pi)) {
+        LOG(WARNING) << "No corresponding PipelinedInfo in socket";
+        return MakeParseError(PARSE_ERROR_TRY_OTHERS);
+    }
+
+    InputResponse* msg = static_cast<InputResponse*>(socket->parsing_context());
```

Comment:
```suggestion
    MySQLInputResponse* msg = static_cast<MySQLInputResponse*>(socket->parsing_context());
```

Reply:
consider replace all InputResponse  with MySQLInputResponse; these redefines are what causing segmentfaults in brpc_redis_unittest